src/libraries/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs

```diff
@@ -9,171 +9,162 @@
 
 namespace System.Net.Sockets.Tests
 {
-    public class DisconnectTest
+    public abstract class Disconnect<T> : SocketTestHelperBase<T> where T : SocketHelperBase, new()
     {
-        private readonly ITestOutputHelper _log;
-
-        public DisconnectTest(ITestOutputHelper output)
-        {
-            _log = TestLogging.GetInstance();
-            Assert.True(Capability.IPv4Support() || Capability.IPv6Support());
-        }
-
-        private static void OnCompleted(object sender, SocketAsyncEventArgs args)
-        {
-            EventWaitHandle handle = (EventWaitHandle)args.UserToken;
-            handle.Set();
-        }
-
-        [Fact]
-        public void InvalidArguments_Throw()
-        {
-            using (Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
-            {
-                AssertExtensions.Throws<ArgumentNullException>("asyncResult", () => s.EndDisconnect(null));
-                AssertExtensions.Throws<ArgumentNullException>("e", () => s.DisconnectAsync(null));
-                AssertExtensions.Throws<ArgumentException>("asyncResult", () => s.EndDisconnect(Task.CompletedTask));
-                s.Dispose();
-                Assert.Throws<ObjectDisposedException>(() => s.Disconnect(true));
-                Assert.Throws<ObjectDisposedException>(() => s.BeginDisconnect(true, null, null));
-                Assert.Throws<ObjectDisposedException>(() => s.EndDisconnect(null));
-                Assert.Throws<ObjectDisposedException>(() => { s.DisconnectAsync(null); });
-            }
-        }
+        protected Disconnect(ITestOutputHelper output) : base(output) { }
 
         [Theory]
         [InlineData(true)]
         [InlineData(false)]
-        [OuterLoop("https://github.com/dotnet/runtime/issues/18406")]
-        public void Disconnect_Success(bool reuseSocket)
+        public async Task Disconnect_Success(bool reuseSocket)
```

Comment:
I think we also need to investigate / test Disconnect behavior during pending operations.

I tried to add a test for that, and discovered an inconsistency across OS-es:

```C#
        [Theory]
        [InlineData(true)]
        [InlineData(false)]
        public async Task Disconnect_DuringPendingReceive_Success(bool reuseSocket)
        {
            IPEndPoint loopback = new IPEndPoint(IPAddress.Loopback, 0);
            using (var server1 = SocketTestServer.SocketTestServerFactory(SocketImplementationType.Async, loopback))
            {
                using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
                {
                    await ConnectAsync(client, server1.EndPoint);
                    Assert.True(client.Connected);

                    var pendingReceive = client.ReceiveAsync(new byte[32], SocketFlags.None);

                    await DisconnectAsync(client, reuseSocket);
                    Assert.False(client.Connected);

                    // On Windows we receive 0 bytes:
                    int bytesReceived = await pendingReceive;
                    Assert.Equal(0, bytesReceived);

                    // On Linux, SocketException will be thrown:
                    // await Assert.ThrowsAsync<SocketException>(() => pendingReceive); // "operation cancelled"
                }
            }
        }
```

Reply:
Huh. How does that compare to doing Shutdown or Close with a pending operation?

BTW, we have existing tests for Shutdown/Close with pending operations, right? Do you know where these are?

Reply:
The only related tests I was able to find are in `SendReceive`: `Receive0ByteReturns_WhenPeerDisconnects`, `TcpReceiveSendGetsCanceledByDispose`, behavior seems to differ.

Reply:
I thought @tmds added some tests related to shutdown and/or close with pending operations, back when he made a big change to cleanup/improve the close logic in general. @tmds?

Reply:
There are CanceledByDispose tests for Receive, Send, Accept, Connect and SendFile.

Reply:
Thanks, that's what I was thinking of. I'll look at those.